Add support for logaddexp(float16) in CUDA and implement its reference

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91869

The reference is implemented so that it generates efficient and
numerically stable triton code.

Fixes https://github.com/pytorch/pytorch/issues/91683

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

lezcano committed:
Add support for logaddexp(float16) in CUDA and implement its reference

The reference is implemented so that it generates efficient and
numerically stable triton code.

Fixes https://github.com/pytorch/pytorch/issues/91683

[ghstack-poisoned]

lezcano commented:
The implementation is slightly better than that we have in CUDA in master I think.

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -1432,6 +1432,20 @@ def le(a: TensorLikeType, b: TensorLikeType) -> TensorLikeType:
+def logaddexp(a: TensorLikeType, b: TensorLikeType) -> TensorLikeType:
+    # Nb. this implementation does nto distribute the gradients evenly when a == b
+    mask = a >= b
+    max_ = torch.where(mask, a, b)
```
I'm not sure 2 masking ops are better than one `abs`, but I don't think it matters much.

lezcano replied:
it's 1 masking op vs 1 abs and a neg. It should be marginally better, but yeah

lezcano replied:
Also, note that nvcc and may be able to fuse the two wheres into just one flow control op.

lezcano committed:
Update on "Add support for logaddexp(float16) in CUDA and implement its reference"


The reference is implemented so that it generates efficient and
numerically stable triton code.

Fixes https://github.com/pytorch/pytorch/issues/91683

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano committed:
Update on "Add support for logaddexp(float16) in CUDA and implement its reference"


The reference is implemented so that it generates efficient and
numerically stable triton code.

Fixes https://github.com/pytorch/pytorch/issues/91683

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano commented:
@pytorchbot merge -f "unrelated rocm error"
